backend/crates/eiffelvis_core/src/graph.rs

```diff
@@ -0,0 +1,108 @@
+use std::cmp::Ordering;
+
+pub trait Key: PartialEq + Eq + Copy + Clone + std::fmt::Debug + std::hash::Hash + Send {}
+pub trait Idx: PartialEq + Eq + Copy + Clone + std::fmt::Debug + std::hash::Hash + Send {}
+impl<T> Idx for T where T: PartialEq + Eq + Copy + Clone + std::fmt::Debug + std::hash::Hash + Send {}
+
+pub trait ValueIndex<Idx>: Sized {
+    type Output;
+
+    fn index(self, idx: Idx) -> Self::Output {
+        self.try_index(idx).unwrap()
+    }
+
+    fn try_index(self, idx: Idx) -> Option<Self::Output>;
+}
+
+pub trait Meta {
+    type Data;
+    type EdgeData;
+    type Idx: Idx;
+    type Key: Key;
+}
+
+pub trait Ref<'a>:
+    Copy
+    + ValueIndex<<Self::Meta as Meta>::Idx, Output = Self::Item>
+    + ValueIndex<<Self::Meta as Meta>::Key, Output = Self::Item>
+where
+    Self::Meta: 'a,
+{
+    type Meta: Meta;
+
+    fn cmp_index(self, lhs: <Self::Meta as Meta>::Idx, rhs: <Self::Meta as Meta>::Idx) -> Ordering;
+
+    type Item: Item<
+        'a,
+        Idx = <Self::Meta as Meta>::Idx,
+        Data = <Self::Meta as Meta>::Data,
+        EdgeData = <Self::Meta as Meta>::EdgeData,
+    >;
+    type ItemIterator: Iterator<Item = Self::Item>;
+    fn items(self) -> Self::ItemIterator;
+}
```

Comment:
Consider adding more specific functions on this so that graphs can implement optimized specifics, for example a ranged iterator.